pkg/pubsubsource/pubsub_test.go

```diff
@@ -0,0 +1,221 @@
+	topic := client.Topic(topicID)
+	exists, err := topic.Exists(ctx)
+	if err != nil {
+		return err
+
+	}
+	if !exists {
+		if _, err = client.CreateTopic(ctx, topicID); err != nil {
+			return err
+		}
+	}
+
+	sub := client.Subscription(subID)
+	exists, err = sub.Exists(ctx)
+	if err != nil {
+		return err
+
+	}
+	if !exists {
+		if _, err = client.CreateSubscription(ctx, subID, pubsub.SubscriptionConfig{Topic: topic}); err != nil {
+			return err
+
+		}
+	}
+	return nil
+}
+
+func TestMain(m *testing.M) {
+	var err error
+	p, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("could not connect to docker ;is it running ? %s", err)
+	}
+	pool = p
+
+	// Check if pubsub container is already running
+	containers, err := pool.Client.ListContainers(docker.ListContainersOptions{})
+	if err != nil {
+		log.Fatalf("could not list containers %s", err)
+	}
+	pubSubRunning := false
+	for _, container := range containers {
+		fmt.Println(container)
+		for _, name := range container.Names {
+			if strings.Contains(name, "google-cloud-cli") {
+				pubSubRunning = true
+				break
+			}
+		}
+		if pubSubRunning {
+			break
+		}
+	}
+
+	if !pubSubRunning {
+		// Start goaws container if not already running
+		opts := dockertest.RunOptions{
+			Repository:   "thekevjames/gcloud-pubsub-emulator",
+			Tag:          "latest",
+			ExposedPorts: []string{"8681"},
+			PortBindings: map[docker.Port][]docker.PortBinding{
+				"8681": {
+					{HostIP: "127.0.0.1", HostPort: "8681"},
+				},
+			},
+		}
+		resource, err = pool.RunWithOptions(&opts)
+		if err != nil {
+			_ = pool.Purge(resource)
+			log.Fatalf("could not start resource %s", err)
+		}
+	}
+
+	err = os.Setenv("PUBSUB_EMULATOR_HOST", PUB_SUB_EMULATOR_HOST)
+	if err != nil {
+		log.Fatalf("error -%s", err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	if err := pool.Retry(func() error {
+		pubsubClient, err = pubsub.NewClient(ctx, Project)
+		if err != nil {
+			log.Fatalf("Failed to create client: %v", err)
+		}
+		return nil
+	}); err != nil {
+		if resource != nil {
+			_ = pool.Purge(resource)
+		}
+		log.Fatalf("could not connect to gcloud pubsub %s", err)
+	}
+	defer pubsubClient.Close()
+	defer cancel()
+	code := m.Run()
+	if resource != nil {
+		if err := pool.Purge(resource); err != nil {
+			log.Fatalf("Couln't purge resource %s", err)
+		}
+	}
+	os.Exit(code)
+}
+
+func TestPubSubSource_Read(t *testing.T) {
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	messageCh := make(chan sourcer.Message, 20)
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD, MAX_OUT_STANDING_MESSAGES)
+	ctx, cancel := context.WithCancel(context.Background())
+	publishChan := make(chan struct{})
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
```

Comment:
Can we move `StartReceiving(ctx)` to inside of NewPubSubSource? The benefit is that we don't need to call StartReceiving every time we New.

Reply:
We can but it  will violate single responsibility principle .

Reply:
Agreed.